plugins/autogems.cpp

```diff
@@ -0,0 +1,293 @@
+                        }
+                    }
+                }
+                
+                add_tasks(piled, workshop);
+            }
+        } else {
+            // Note which gem types have already been ordered to be cut.
+            for (auto j = workshop->jobs.begin(); j != workshop->jobs.end(); ++j) {
+                auto job = *j;
+                if (job->job_type == df::job_type::CutGems) {
+                    available[job->mat_index] -= job->flags.bits.repeat? 100: 1;
+                }
+            }
+            
+            if (workshop->jobs.size() <= MAX_WORKSHOP_JOBS) {
+                unlinked.insert(workshop);
+            }
+        }
+    }
+    
+    if (unlinked.size() > 0) {
+        // Count how many gems of each type are available to be cut.
+        // Gems in stockpiles linked to specific workshops don't count.
+        auto gems = world->items.other[items_other_id::ROUGH];
+        for (auto g = gems.begin(); g != gems.end(); ++g) {
+            auto item = *g;
+            // ROUGH also includes raw glass; the INORGANIC check filters that out.
+            if (item->getMaterial() == builtin_mats::INORGANIC && !stockpiled.count(item->id)) {
+                available[item->getMaterialIndex()] += 1;
+            }
+        }
+        
+        for (auto w = unlinked.begin(); w != unlinked.end(); ++w) {
+            add_tasks(available, *w);
+        }
+    }
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out) {
+    if (running && (world->frame_counter - last_frame_count >= DELTA_TICKS)) {
+        last_frame_count = world->frame_counter;
+        create_jobs();
+    }
+    
+    return CR_OK;
+}
+
+
+/*
+ * Interface hooks
+ */
+struct autogem_hook : public df::viewscreen_dwarfmodest {
+    typedef df::viewscreen_dwarfmodest interpose_base;
+    
+    bool in_menu() {
+        // Determines whether we're looking at the Workshop Orders screen.
+        return ui->main.mode == ui_sidebar_mode::OrdersWorkshop;
+    }
+    
+    bool handleInput(std::set<df::interface_key> *input) {
+        if (!in_menu()) {
+            return false;
+        }
+        
+        if (input->count(interface_key::CUSTOM_G)) {
+            // Toggle whether gems are auto-cut for this fort.
+            auto config = World::GetPersistentData(CONFIG_KEY, NULL);
+            if (config.isValid()) {
+                config.ival(0) = running;
+            }
+            
+            running = !running;
+            return true;
+        }
+        
+        return false;
+    }
+    
+    DEFINE_VMETHOD_INTERPOSE(void, feed, (std::set<df::interface_key> *input)) {
+        if (!handleInput(input)) {
+            INTERPOSE_NEXT(feed)(input);
+        }
+    }
+    
+    DEFINE_VMETHOD_INTERPOSE(void, render, ()) {
+        INTERPOSE_NEXT(render)();
+        if (in_menu()) {
+            auto dims = Gui::getDwarfmodeViewDims();
+            int x = dims.menu_x1 + 1;
+            int y = dims.y1 + 12;
```

Comment:
Just curious - do you know of an easy way to get those options to show up (for testing purposes)?